hal-core/src/interrupt.rs

```diff
@@ -0,0 +1,125 @@
+/// An interrupt controller for a platform.
+pub trait Control {
+    type Error;
+    /// The type of this platform's interrupt vector.
+    ///
+    /// TODO(eliza): This is *probably* always a `u8`, do we really need to have
+    /// this?
+    type Vector;
+
+    /// Disable interrupts.
+    unsafe fn disable_irq(&mut self);
```

Comment:
why is this and `enable_irq` `unsafe`? because irqs (and masking) are global mutable state? i'm perfectly fine with this for my own terrible reasoning i'm just curious

Reply:
IIRC if you `sti` twice in a row it does a general protection fault or something. maybe i made that up though?

Reply:
upon further consideration it is probably the platform impl’s job to “not do that”; these should probably be safe

Reply:
i also considered having geese take `&mut self` even though they are probably always going to be a single instruction that doesn’t actually _mutate_ the receiver...but maybe the platform impl wants to do bookkeeping as well? idk

Reply:
i can definitely imagine impls that would want to do bookkeeping, yes

Reply:
there should probably also be a `fn mask(Interrupt<Ctrl=Self>)` and a corresponding `unmask`; there could be a drop guard API implemented on top of those as well